### DISCUSSION TITLE
Transformer Test Case Problems

### DISCUSSION THREAD

**Discussion by Boshen** [oxc-project/oxc#3251] [General]
This discussion records all the problems we encountered while working on the transformer.

Each answer is for a specific problem and its matching test cases.

cc @overlookmotel @Dunqing

**Comment by Dunqing** [comment]
## [scope.hasOwnBinding()](https://github.com/babel/babel/blob/4bd1b2c2f1bb3f702cfcb50448736e33c7000128/packages/babel-traverse/src/scope/index.ts#L1303-L1305)

Call from [here](https://github.com/babel/babel/blob/4bd1b2c2f1bb3f702cfcb50448736e33c7000128/packages/babel-plugin-transform-typescript/src/enum.ts#L145)

### babel-plugin-transform-typescript 
* imports/enum-value/input.ts

**Reply by Dunqing** [reply]
We need to fully support semantic’s scopes and symbols in TraverseCtx. but we don’t care about type-only symbols

**Reply by overlookmotel** [reply]
This is now doable via the APIs added in #3314. I don't think we need a special method/abstraction for this one - we can interface with the scopes tree directly.

**Reply by overlookmotel** [reply]
I tried to tackle this one, but am finding the logic hard to follow.

It doesn't help that Babel's logic for when to take a binding in upper scope vs a previous enum variant doesn't seem to be consistent either: [Babel REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=false&spec=false&loose=false&code_lz=JYWwDg9gTgLgBAbzgQQDRwMLoKJwL5wBmUEIcA5ADbABG5A3AFCMD0LcAEhAO5zAAmAUwB2MYIWCCoAZzgBDKILiLpESgDdB_ONMGCQs4MIDGEYdODSYImMzZwABgBE4AXkwO4AV12yaR_iMAcz5hbzAwKR1TSMYRLzJsYQTERjhMNxRUNLgXdyxGPDt2BwAxTOxPH0FZB0q4dQVgOVE4CEI4eJA45MTegCZU9Nx3AEZs9PL3bGyioA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=&prettier=true&targets=&version=7.24.5&externalPlugins=%40babel%2Fplugin-transform-typescript%407.24.5&assumptions=%7B%7D)

@Dunqing I think this one might be easier for you handle, as you wrote the transform originally. Would you be willing to give it a go?

The `scope.hasOwnBinding(name)` equivalent is `ctx.scopes().has_binding(scope_id, name)`.

**Reply by Dunqing** [reply]
@overlookmotel I am sorry about that! I forgot to add TODO comment to what we need to check binding. Actually the logic of our implementation is same as that of Babel. I have figured out this case and added some comments in #3387. You can check it out!

**Reply by Dunqing** [reply]
Another problem. I want to refactor this based on `Traverse`. Because I need to use `ctx.scopes.get_binding()` to check something.

https://github.com/oxc-project/oxc/blob/3c55da6e015963ac6d04c32f22e7495d7efdd8b3/crates/oxc_transformer/src/typescript/enum.rs#L511-L552

But here we can't use `traverse_mut`, I just have an `Expression`. Seems to make it possible, we need to expose the walk::* function? Is there any better way to do this?

**Reply by overlookmotel** [reply]
I actually don't know if it safe to expose `walk_*` functions or not. At present the scheme is very careful not to create references during traversal, but here you start with a `&mut Expression`, so that could *potentially* cause UB if we then called `walk_expression` with it.

I do not understand Rust's memory model well enough to say either way whether this is safe. Maybe it is, but maybe not. I will post a question on the Rust forums. Some super-experienced people (like Ralf Jung) hang out there, and maybe we can get their advice.

But for now I would like to be cautious. Can we avoid exposing `walk_*` until we know it's safe?

For now I think you could achieve what you are looking for by making `&mut scopes` a field on `IdentifierReferenceRename`, and get therefore make `scopes` accessible within its `visit_expression` visitor. Would that work?

**Reply by Dunqing** [reply]
> But for now I would like to be cautious. Can we avoid exposing walk_* until we know it's safe?

We can leave it for now.

> For now I think you could achieve what you are looking for by making &mut scopes a field on IdentifierReferenceRename, and get therefore make scopes accessible within its visit_expression visitor. Would that work?

That doesn't work. The `Expression` is `FunctionExpression`. That means we need to mutate scopes. 
Look at the example
```ts
enum A {
  a = 1,
  b = function b() { // is expression
        let a = 11;
        a // check binding here
  }
}
```

**Reply by overlookmotel** [reply]
Hmm I see the problem. I think you *can* do this with `VisitMut`, but you'd need to add visitors to keep track of what scope you're in. e.g. `visit_function` would push the `scope_id` of the function to a stack, so you can then read it back later. But this means re-implementing all the scope-tracking functionality that `Traverse` already has. This is possible, but a huge pain.

**Reply by overlookmotel** [reply]
Actually... could something like this work?

1. Add a flag `is_in_enum` to `TypeScriptEnum`.
2. Set that flag to `true` when entering an enum, and back to `false` when exiting it.
3. Add a visitor for `enter_identifier_reference` to `TypeScriptEnum`.
4. In `enter_identifier_reference`, it checks if `is_in_enum` is `true`, and if it is then it reads `ctx.scopes()` and can decide whether to replace the identifier or not.

i.e. Do everything in one `Traverse` pass, instead of traverse-within-traverse.

I *think* this could work? But maybe I missed something?

**Reply by Dunqing** [reply]
Sounds good. I've thought it through. The only problem is `previous_enum_members`. In each transform enum member. We will cache the value of enum member to `previous_enum_members`. The reason for doing this is because the initializer of an enum member can use the enum members that have already been defined before it. So we may not solve this [case](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=KYOwrgtgBAglDeAoKUCGUC8UAMAaZUARpmlANRQCM-KAxiQCY1QMkAs-AvkA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ctypescript&prettier=false&targets=&version=7.24.5&externalPlugins=&assumptions=%7B%7D).

**Reply by overlookmotel** [reply]
I guess it's possible to have an enum within an enum.

```typescript
enum A {
  a = 1,
  b = function() {
    enum B {
      c = a,
      d = x,
    }
  }
}
```

so instead of a single flag `is_in_enum`, would need a stack of "current enums", to support this bizarre use case.

> So we may not solve this [case](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=KYOwrgtgBAglDeAoKUCGUC8UAMAaZUARpmlANRQCM-KAxiQCY1QMkAs-AvkA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ctypescript&prettier=false&targets=&version=7.24.5&externalPlugins=&assumptions=%7B%7D).

I don't see the problem with that one. What's the catch? If no binding found in upper scope, it defaults to resolving `foo` as `Enum.foo`. So, in that example, is it a problem that `d` is used before it's defined?

**Reply by Dunqing** [reply]
Here is the entry point for the transform enum.

https://github.com/oxc-project/oxc/blob/3c55da6e015963ac6d04c32f22e7495d7efdd8b3/crates/oxc_transformer/src/typescript/mod.rs#L140-L147
This means we start traversing the new statement only after transforming the whole enum.

I've thought it over carefully, and it's getting a bit complicated. Because the value of an enum member can be any `Expression`. If we want to modify the identifier reference after transforming the enum, we may encounter many edge-case

**Comment by Dunqing** [comment]
## [scope.generateUid()](https://github.com/babel/babel/blob/4bd1b2c2f1bb3f702cfcb50448736e33c7000128/packages/babel-traverse/src/scope/index.ts#L500-L522)

Call from [here](https://github.com/babel/babel/blob/4bd1b2c2f1bb3f702cfcb50448736e33c7000128/packages/babel-plugin-transform-typescript/src/namespace.ts#L127-L128)

### babel-plugin-transform-typescript 
* namespace/canonical/input.ts
* namespace/contentious-names/input.ts
* namespace/empty-removed/input.ts
* namespace/module-nested/input.ts
* namespace/module-nested-export/input.ts
* namespace/multiple/input.ts
* namespace/mutable-fail/input.ts
* namespace/namespace-flag/input.ts
* namespace/namespace-nested-module/input.ts
* namespace/nested/input.ts
* namespace/nested-destructuring/input.ts
* namespace/nested-namespace/input.ts
* namespace/nested-shorthand-export/input.ts
* namespace/same-name/input.ts

### babel-plugin-transform-react-jsx 
* autoImport/complicated-scope-module/input.js

**Reply by overlookmotel** [reply]
Question: Instead of `generateUid()`, could we provide `ctx.generateBinding()` which would work like:

* `ctx.generateBinding()` returns an `Atom` which you can then use in generated code.
* When traversal exits the parent function/program, *then* it adds `var _temp;` declaration to AST.
* At that point, it knows all the existing bindings in the scope, so can choose a var name which doesn't clash.
* At the very end of traversal, after `exit_program`, it then "fixes up" all the places where you used the `Atom` that `generateBinding()` returned.

The potential problems I can see with this are:

1. The `Atom` returned from `generateBinding()` will have a name which may change later (if required because of another binding with this name discovered later).
2. Only allows creating a binding (`var _temp;`), not a binding with initialization (`var _temp = blah;`).
3. While you continue traversing, you can't see the `var _temp;` statement, as it's only added later on.

But I'm not sure if in practice any of these problems apply for React and TS transforms?

**Reply by overlookmotel** [reply]
We could add bindings at end of the block, and it'd still work. e.g. for down-leveling array destructuring:

```js
function foo() {
  if (bar) {
    const [x, y] = getArray();
    run(x, y);
  }
}
```

could become:

```js
function foo() {
  if (bar) {
    _temp = getArray();
    const x = _temp[0], y = _temp[1]; // Simplified
    run(x, y);
  }
  var _temp;
}
```

This is valid output and would be easier to implement, because you don't alter the indexes of the statements in the function, just add a new one to the end.

But I assume that's not an option for us, as it won't match Babel's output so the tests would fail?

**Reply by Dunqing** [reply]
> Question: Instead of `generateUid()`, could we provide `ctx.generateBinding()` which would work like:
> 
> * `ctx.generateBinding()` returns an `Atom` which you can then use in generated code.
> * When traversal exits the parent function/program, *then* it adds `var _temp;` declaration to AST.
> * At that point, it knows all the existing bindings in the scope, so can choose a var name which doesn't clash.
> * At the very end of traversal, after `exit_program`, it then "fixes up" all the places where you used the `Atom` that `generateBinding()` returned.

If I understand correctly. Does the implementation of `ctx.generateBinding()` also need to use `generateUid()`？ The `ctx.generateBinding()` sounds good. But I am not sure if the behavior weather matches `Babel`.  I think we can first support `generateUid()` and then confirm whether we need to provide `ctx.generateBinding()` based on the use case.

> But I assume that's not an option for us, as it won't match Babel's output so the tests would fail?

Yes.

**Reply by overlookmotel** [reply]
Thanks for feedback. I'm going through some of the list above and trying to figure out what we need. Am trying to simplify the requirements, but not having much success!

**Reply by magic-akari** [reply]
> But I assume that's not an option for us, as it won't match Babel's output so the tests would fail?

Babel's transformation meets its own needs.
If we have a transformation that suits us better, perhaps we can choose not to align with them.

> This is valid output and would be easier to implement, because you don't alter the indexes of the statements in the function, just add a new one to the end.

There is a similar example, which is to insert an import helper. The import statement can be placed at the end of the file, and that's fine, but it's important that it does not conflict with any existing variables in JavaScript.

**Reply by magic-akari** [reply]
In Test262, there are some executable test cases.
We can test whether those codes can still output the correct results after being transformed.
This is a good supplement to the Babel test cases.

**Reply by overlookmotel** [reply]
Good points. It is frustrating that we have to at present align with Babel's output. What if we find more optimal solutions than Babel? We skip them? Or we make PRs on Babel to improve it too??

On the other hand, I do really like that we can just "borrow" all Babel's tests, and don't have to write them ourselves. Writing tests is no-one's idea of a good time, and Babel is heavily battle-tested already, so not diverging from it is a good recipe for moving fast while avoiding bugs.

Perhaps a good strategy is to align with Babel's output for now, but make notes as to where we think we can improve on it. Then, once we have all the existing tests passing, revisit them and optimize.

But adding the Test262 exec test cases sounds like a good idea regardless to get more test coverage.

**Reply by overlookmotel** [reply]
#3314 added all the primitives required to support this. I will build an abstraction around them to produce a `generateUid`-type API next week.

**Reply by overlookmotel** [reply]
Please see #3394 and #3395.

**Reply by overlookmotel** [reply]
@Dunqing As you mentioned in https://github.com/oxc-project/oxc/pull/3395#issuecomment-2128520415, it sounds like you're going to tackle the rest of these, now that `TraverseCtx::generate_uid` is implemented.

But please can you leave the React JSX one for me? I would like to get some practice on the transformer, and it would be easier for me to do that on a JSX transform (where I understand what the transform is doing), rather than TypeScript ones (where I don't understand, because I don't know TS).

**Reply by Dunqing** [reply]
Sure, enjoy it.

**Reply by overlookmotel** [reply]
#3431 solves the React JSX one.

**Comment by Dunqing** [comment]
## Complicated cases
### babel-plugin-transform-typescript 
* enum/mix-references/input.ts
* enum/ts5.0-const-foldable/input.ts

These tests require the support of these API

-  [scope.hasOwnBinding](https://github.com/babel/babel/blob/4bd1b2c2f1bb3f702cfcb50448736e33c7000128/packages/babel-traverse/src/scope/index.ts#L1303-L1305) - call from [here](https://github.com/babel/babel/blob/4bd1b2c2f1bb3f702cfcb50448736e33c7000128/packages/babel-plugin-transform-typescript/src/enum.ts#L145)
- [scope.getBindingIdentifier()](https://github.com/babel/babel/blob/4bd1b2c2f1bb3f702cfcb50448736e33c7000128/packages/babel-traverse/src/scope/index.ts#L1293-L1295) - call from [here](https://github.com/babel/babel/blob/4bd1b2c2f1bb3f702cfcb50448736e33c7000128/packages/babel-plugin-transform-typescript/src/enum.ts#L158) 
- [path.resolve()](https://github.com/babel/babel/blob/4bd1b2c2f1bb3f702cfcb50448736e33c7000128/packages/babel-traverse/src/path/introspection.ts#L534-L543) - call from [here](https://github.com/babel/babel/blob/4bd1b2c2f1bb3f702cfcb50448736e33c7000128/packages/babel-plugin-transform-typescript/src/enum.ts#L327)

**Reply by overlookmotel** [reply]
`scope.hasOwnBinding` and `scope.getBindingIdentifier` equivalents are now available via the APIs added in #3314.

Not sure about `path.resolve` - I've not investigated what it does yet.

**Reply by Dunqing** [reply]
What is the `scope.getBindingIdentifier ` equivalent api?

`path.resolve()` is to find the defined node based on the current node. For example, if the current node is an `IdentifierReference`. The return value of `path.resolve` is a `BindingIdentifier`

```ts
let a = 0;
    ^
    The result of path.resolve() 
console.log(a) // call path.resolve() here
```

**Comment by Dunqing** [comment]
## ScopeFlags

### babel-plugin-transform-react-jsx-self
* react-source/disable-with-super/input.js

### babel-plugin-transform-react-jsx-development
* cross-platform/within-derived-classes-constructor/input.js

These tests need to know what the current node's ScopeFlags are

**Reply by overlookmotel** [reply]
This one has been addressed in #3229. `ctx.scope()` gives you `ScopeFlags` for the current node.

Please see notes in that PR about inheritance of flags from parent scope.

**Reply by Dunqing** [reply]
Thank you! These tests were done.

**Comment by Dunqing** [comment]
### babel-plugin-transform-react-jsx-development
* cross-platform/within-ts-module-block/input.ts

This test needs to enable typescript in codegen and turn off the typescript plugin

**Reply by Boshen** [reply]
I'll follow up on this.

**Reply by Boshen** [reply]
Need to add an option to disable typescript plugin.

**Comment by Dunqing** [comment]
We need to port [registerDeclaration](https://github.com/babel/babel/blob/ab465cc0ab11196bcc8ef48068984e16193d8198/packages/babel-traverse/src/scope/index.ts#L746) api. In the namespace, we will transform the `namespace {}` to `function() {}`. But, currently, the new function has no `scope_id`. We need to create a scope and bind it to a new function. Currently, block this case `namespace/empty-removed/input.ts`

**Reply by Dunqing** [reply]
@overlookmotel cc

**Reply by overlookmotel** [reply]
If I understand how `namespace {}` works, `TSModuleBlock` represents the `{}` part and it has a `scope_id`.

The transform removes the `TSModuleBlock`, and replaces with a `function`. So can we "steal" the `scope_id` from the old `TSModuleBlock` and make it the `scope_id` for the new function instead?

**Reply by Dunqing** [reply]
Yes, It's a workaround. I just wanted to let you know that we can do it for now. Their scope flags are different. We may also need to provide a method to change it.

**Reply by Dunqing** [reply]
I have done in #3459 #3460. All namespace test cases are done!

**Reply by overlookmotel** [reply]
> All namespace test cases are done!

🍾 Bravo Dunqing!

I will review those 2 PRs later today.

Yes, we definitely need to add new APIs for manipulating scopes, symbols + references.

In this particular case, we should be able to alter the scope flags for the old `TSModuleBlock` scope via `ctx.scopes_mut()`.

**Comment by overlookmotel** [comment]
Do we need to transform `UsingDeclaration` for this milestone?

As #3482 points out, it's not exclusively TS syntax, but as it's currently a [stage 3 proposal](https://github.com/tc39/proposal-explicit-resource-management) it will likely need to be transpiled to run in browsers/NodeJS. caniuse.com doesn't list it, which I assume means it's not implemented in browsers yet.

**Reply by Boshen** [reply]
It's not part of the current milestone.
